File: backend/app/services/simulation.py
from netra import Netra
from netra.simulation import BaseTask, TaskResult
from typing import Optional
from uuid import uuid4
from agent import get_response

class LoanAgentTask(BaseTask):
    
    def run(self, message: str, session_id: Optional[str] = None) -> TaskResult:
        thread_id = uuid4().hex if not session_id else session_id
    
        Netra.set_session_id(thread_id)

        # Get response from the agent
        try:
            response = get_response(message, thread_id)
            final_message = response
        except Exception as e:
            
            final_message = f"Error: {str(e)}"

        return TaskResult(
            message=final_message,
            session_id=thread_id
        )
    
def run_simulation(dataset_id: str) -> None:
    Netra.simulation.run_simulation( #type: ignore
        name="Loan Agent Simulation",
        dataset_id=dataset_id,
        task=LoanAgentTask()
    )

File: backend-ts/src/services/simulation.ts
import { Netra, BaseTask, type TaskResult } from "netra-sdk";
import { getResponse } from "../agent/index.js";
import { randomUUID } from "node:crypto";

class LoanAgentTask extends BaseTask {
  async run(message: string, sessionId?: string | null): Promise<TaskResult> {
    const threadId = sessionId || randomUUID().replace(/-/g, "");

    Netra.setSessionId(threadId);

    let finalMessage: string;
    // Get response from the agent
    try {
      const response = await getResponse(message, threadId);
      finalMessage = response;
    } catch (e: any) {
      finalMessage = `Error: ${e.message}`;
    }

    return {
      message: finalMessage,
      sessionId: threadId,
    };
  }
}

export const runSimulation = async (datasetId: string) => {
  return await Netra.simulation.runSimulation({
    name: "Loan Agent Simulation",
    datasetId: datasetId,
    task: new LoanAgentTask(),
  });
};
